### PR TITLE
Correctly create custom analyzer in Elasticsearch

### LINKED ISSUE #17012: Custom analyzers in Elasticsearch no longer work
### Describe the bug
When specifying a custom analyzer in Elasticsearch as explained here https://docs.orchardcore.net/en/latest/reference/modules/Elasticsearch/#elasticsearch-analyzers it does create that custom analyzer in the Elasticsearch service.

### Orchard Core version
2.x

### To Reproduce

Configure custom analyzer like this:
```
"OrchardCore_Elasticsearch": {
      // Other settings
      "Analyzers": {
        "standard": {
          "type": "standard"
        },
        "keyword": {
          "type": "keyword"
        },
        "whitespace": {
          "type": "whitespace"
        },
        "knowledgebase_analyzer": {
          "type": "custom",
          "tokenizer": "standard",
          "filter": [
            "lowercase",
            "stop"
          ],
          "char_filter": [
            "html_strip"
          ]
        }
      }
    }
```
Create an index using the custom `knowledgebase_analyzer`. Now try to search for anything you'll get an error 

> [multi_match] analyzer [knowledgebase_analyzer] not found

This issue was caused by PR #16843

### PR DESCRIPTION
**Pull request by MikeAlhayek**
Fix #17012 

/cc: @denispetrische

### CODE REVIEW

**Comment by denispetrische** [comment]
Let me see.. This pr will disable adding custom token filters, but in the version in main branch that perfectly works.

```json 
"OrchardCore_Elasticsearch": {
  "ConnectionType": "SingleNodeConnectionPool",
  "Url": "http://localhost",
  "Ports": [ 9200 ],
  "Analyzers": {
    "default": {
      "type": "standard",
      "filter": [
        "lowercase",
        "english_stop",
        "english_stemmer"
      ]
    },
    "custom_analyzer": {
      "type": "standard",
      "filter": [
        "lowercase",
        "english_stop",
        "english_stemmer",
        "english_synonyms"
      ]
    }
  },
  "Filter": {
    "english_stop": {
      "type": "stop",
      "stopwords": "_english_"
    },
    "english_stemmer": {
      "type": "stemmer",
      "language": "english"
    },
    "english_synonyms": {
      "type": "synonym_graph",
      "synonyms": ["page,cage"]
    }
  }
}
```

In this pr version analyzer is created without token filters if specified like in configuration above.
image

if to use your configuration like in https://github.com/OrchardCMS/OrchardCore/issues/17012
yeah - it works, but if I only add my custom token filter - elastic index is not created

```json
"OrchardCore_Elasticsearch": {
  "ConnectionType": "SingleNodeConnectionPool",
  "Url": "http://localhost",
  "Ports": [ 9200 ],
  "Analyzers": {
    "standard": {
      "type": "standard"
    },
    "keyword": {
      "type": "keyword"
    },
    "whitespace": {
      "type": "whitespace"
    },
    "knowledgebase_analyzer": {
      "type": "custom",
      "tokenizer": "standard",
      "filter": [
        "lowercase",
        "stop",
        "english_synonyms"
      ],
      "char_filter": [
        "html_strip"
      ]
    }
  },
  "Filter": {
    "english_stop": {
      "type": "stop",
      "stopwords": "_english_"
    },
    "english_stemmer": {
      "type": "stemmer",
      "language": "english"
    },
    "english_synonyms": {
      "type": "synonym_graph",
      "synonyms": [ "page,cage" ]
    }
  }
}
```

I need to understand what exactly no longer work, because everything in main branch seems to work for me.

Anyway we need to be able to add custom token filters that listed in the "Filter" section to any analyzer

**Comment by denispetrische** [comment]
I think I found the reason, I will sugest changes soon

**Comment by MikeAlhayek** [comment]
@denispetrische can you checkout this branch and test out filters? Note that I renamed "Filter" to "TokenFilters"

so in your appsettings use "TokenFilters" instead of "Filter" to define the custom token-filters.

**Comment by denispetrische** [comment]
> @denispetrische can you checkout this branch and test out filters? Note that I renamed "Filter" to "TokenFilters"
> 
> so in your appsettings use "TokenFilters" instead of "Filter" to define the custom token-filters.

I tested locally "fix logs message" and made my own pull request based on it https://github.com/OrchardCMS/OrchardCore/pull/17019. Check this. This works with every analyzer and custom token filters.

I don't think that's a good idea to rename "Filter" section to "TokenFilters" because it is how it called in elastic terminology,

**Comment by MikeAlhayek** [comment]
@denispetrische 

> I tested locally "fix logs message" 

Did you test this branch?

> made my own pull request based

It better to suggest a fix here instead of create a new PR to avoid redundant work.

**Comment by denispetrische** [comment]
> Did you test this branch?
- Yeah, it still doesn't work when I create my own analyzer. 

Ok, I'll suggest them here

File: src/OrchardCore/OrchardCore.Search.Elasticsearch.Core/Services/ElasticIndexManager.cs [hunk 1]
```diff
@@ -597,45 +532,15 @@ private TokenFiltersDescriptor GetTokenFilterDescriptor(Dictionary<string, JsonO
...
-                            var propertyValue = JsonSerializer.Deserialize<string>(filterProperty.Value);
-                            property.SetValue(tokenFilterBuildingInfo.TokenFilter, new StopWords(propertyValue));
-                        }
-
-                        tokenFilterBuildingInfo.AddTokenFilter(descriptor, tokenFilterBuildingInfo.TokenFilter, filterName);
-                        _tokenFilterNames.Add(filterName);
-
-                        continue;
-                    }
+                    PopulateValue(tokenFilterBuildingInfo.TokenFilter, property, filterProperty.Value);
```
Comment (denispetrische) [inline src/OrchardCore/OrchardCore.Search.Elasticsearch.Core/Services/ElasticIndexManager.cs:536]:
```suggestion
private bool IsAnalyzerContainsCustomTokenFilters(IAnalyzer analyzer)
{
    if (analyzer.GetType() != typeof(CustomAnalyzer))
    {
        return false;
    }

    var filterProperty = analyzer.GetType().GetProperty("Filter");

    if (filterProperty == null)
    {
        return false;
    }

    var filterNames = filterProperty.GetValue(analyzer) as IEnumerable<string>;

    if (!filterNames.Any())
    {
        return false;
    }

    foreach (var filterName in filterNames)
    {
        if (_customTokenFilterNames.Contains(filterName))
        {
            return true;
        }
    }

    return false;
}
```

Comment (denispetrische) [reply src/OrchardCore/OrchardCore.Search.Elasticsearch.Core/Services/ElasticIndexManager.cs:536]: And to insert this method to ElasticIndexManager.cs

File: src/OrchardCore/OrchardCore.Search.Elasticsearch.Core/Services/ElasticIndexManager.cs [hunk 2]
```diff
@@ -439,35 +391,23 @@ public async Task<bool> CreateIndexAsync(ElasticIndexSettings elasticIndexSettin
         ? null
         : elasticIndexSettings.AnalyzerName) ?? ElasticsearchConstants.DefaultAnalyzer;
 
-        if (_elasticSearchOptions.Analyzers.TryGetValue(analyzerName, out var analyzerProperties))
+        if (_elasticSearchOptions.Analyzers is not null && _elasticSearchOptions.Analyzers.TryGetValue(analyzerName, out var analyzerProperties))
```
Comment (denispetrische) [inline src/OrchardCore/OrchardCore.Search.Elasticsearch.Core/Services/ElasticIndexManager.cs:394]:
```suggestion
        if (_elasticSearchOptions.Filter is not null && _elasticSearchOptions.Filter.Count > 0)
{
    var tokenFiltersDescriptor = GetTokenFilterDescriptor(_elasticSearchOptions.Filter);
    analysisDescriptor.TokenFilters(f => tokenFiltersDescriptor);
}

if (_elasticSearchOptions.Analyzers.TryGetValue(analyzerName, out var analyzerProperties))
{
    var analyzer = CreateAnalyzer(analyzerProperties);

    if (IsAnalyzerContainsCustomTokenFilters(analyzer))
    {
        var existingCustomTokenFilters = (analyzer as CustomAnalyzer).Filter.ToList();
        indexSettingsDescriptor.Analysis(descriptor => analysisDescriptor.Analyzers(a => a.Custom(analyzerName, c => c.Tokenizer("standard").Filters(existingCustomTokenFilters))));
    }
    else
    {
        indexSettingsDescriptor.Analysis(descriptor => analysisDescriptor.Analyzers(a => a.UserDefined(analyzerName, analyzer)));
    }
}
```

Comment (denispetrische) [reply src/OrchardCore/OrchardCore.Search.Elasticsearch.Core/Services/ElasticIndexManager.cs:394]: Instead of string 394 - 404

**Review by denispetrische — CHANGES_REQUESTED** [review]

**Comment by MikeAlhayek** [comment]
@denispetrische I have cleaned up the code quite bit. Please revisit your suggestions and see if you still have any suggestions. The code with it's current state is much cleaner.

I just tested creating a new token-filter and use it in my custom analyzer.

This is what I specified in the appsettings:

```
  "Analyzers": {
    "standard": {
      "type": "standard"
    },
    "keyword": {
      "type": "keyword"
    },
    "whitespace": {
      "type": "whitespace"
    },
    "knowledgebase_analyzer": {
      "type": "custom",
      "tokenizer": "standard",
      "filter": [
        "lowercase",
        "stop",
        "english_stop"
      ],
      "char_filter": [
        "html_strip"
      ]
    }
  },
  "TokenFilters": {
    "english_stop": {
      "type": "stop",
      "stopwords": "_english_"
    }
  }
```

Then created an index called `tf_test` using my custom `knowledgebase_analyzer` analyzer. When I directly query the index settings using Kabana I get the expected structure 
```
{
  "ptest_blog2_tf_test": {
    "settings": {
      "index": {
        "routing": {
          "allocation": {
            "include": {
              "_tier_preference": "data_content"
            }
          }
        },
        "number_of_shards": "1",
        "provided_name": "ptest_blog2_tf_test",
        "creation_date": "1731606374163",
        "analysis": {
          "filter": {
            "english_stop": {
              "type": "stop",
              "stopwords": "_english_"
            }
          },
          "analyzer": {
            "knowledgebase_analyzer": {
              "filter": [
                "lowercase",
                "stop",
                "english_stop"
              ],
              "char_filter": [
                "html_strip"
              ],
              "type": "custom",
              "tokenizer": "standard"
            }
          }
        },
        "number_of_replicas": "1",
        "uuid": "lDrk5TeTTaiD9lU27YUumg",
        "version": {
          "created": "8060199"
        }
      }
    }
  }
}
```

So the changes work with no problem.

**Comment by denispetrische** [comment]
> ```json
> "OrchardCore_Elasticsearch": {
>   "ConnectionType": "SingleNodeConnectionPool",
>   "Url": "http://localhost",
>   "Ports": [ 9200 ],
>   "Analyzers": {
>     "standard": {
>       "type": "standard"
>     },
>     "keyword": {
>       "type": "keyword"
>     },
>     "whitespace": {
>       "type": "whitespace"
>     },
>     "knowledgebase_analyzer": {
>       "type": "custom",
>       "tokenizer": "standard",
>       "filter": [
>         "lowercase",
>         "stop",
>         "english_synonyms"
>       ],
>       "char_filter": [
>         "html_strip"
>       ]
>     }
>   },
>   "Filter": {
>     "english_stop": {
>       "type": "stop",
>       "stopwords": "_english_"
>     },
>     "english_stemmer": {
>       "type": "stemmer",
>       "language": "english"
>     },
>     "english_synonyms": {
>       "type": "synonym_graph",
>       "synonyms": [ "page,cage" ]
>     }
>   }
> }
> ```

Yes, now that works for me as well. And the code quality is mush better now. Good job)

**Review by denispetrische — APPROVED** [review]

**Pull request merged**